pytests/operators/test_stateful.py

```diff
@@ -284,8 +284,57 @@ def test_stateful_snapshots_discard_per_key(recovery_config):
     ]
 
     out.clear()
-    run_main(flow, epoch_interval=ZERO_TD, recovery_config=recovery_config)
+    run_main(flow, epoch_interval=ZERO_TD, recovery_config=recovery_config_immediate)
     assert out == [
         ("a", (None, "a3")),
         ("b", ("b2", "b3")),
     ]
+
+
+def test_stateful_recovers_older_snapshots(recovery_config_immediate):
+    inp = [
+        ("a", "a1"),
+        TestingSource.ABORT(),
+        ("b", "b1"),
+        TestingSource.ABORT(),
+        ("c", "c1"),
+        ("b", "DISCARD"),
+        TestingSource.ABORT(),
+        ("a", "a2"),
+        ("b", "b2"),
+        ("c", "c2"),
+    ]
+    out = []
+
+    flow = Dataflow("test_df")
+    s = op.input("inp", flow, TestingSource(inp))
+    s = op.stateful("stateful", s, KeepLastLogic)
+    op.output("out", s, TestingSink(out))
+
+    run_main(flow, epoch_interval=ZERO_TD, recovery_config=recovery_config_immediate)
+    assert out == [
+        ("a", (None, "a1")),
+    ]
+
+    out.clear()
+    run_main(flow, epoch_interval=ZERO_TD, recovery_config=recovery_config_immediate)
+    assert out == [
+        ("b", (None, "b1")),
+    ]
+
+    out.clear()
+    run_main(flow, epoch_interval=ZERO_TD, recovery_config=recovery_config_immediate)
+    assert out == [
+        ("b", ("b1", "b1")),
```

Comment:
I did add this as a regression test to check for a bug I had during the implementation of the fixes, but forgot to address this. Looking at it now, it seems to be just a bug in `TestingSource`, that only shows up if the dataflow starts with a triggered `ABORT` and aborts again, so it never showed because we never used the `ABORT` more than once in other tests.

Given that our state is the index into the list, here's what happens.
First run:
- `next_batch` => before emitting, index (state) goes from `0` to `1` (`self._start_idx += len(batch)`)
- `snapshot` => returns `1`
- `next_batch` => We get the first `ABORT`, returns an empty list, and we don't update the state
- `snapshot` => returns `1`
- `next_batch` => raises `ABORT` and ends the first run.

All is fine up until now. Second run, resume_state is `1`:

- `next_batch` => We get the first, but triggered, `ABORT`, so the flow keeps going, but again we don't update the state here (here's the bug).
- `snapshot` => Still returns `1` (but it should be `2` at this point)
- `next_batch` => We get the item with index 2 `('b', 'b1')`
- `snapshot` => Returns `2` (but it should be `3` at this point)
- `next_batch` => We get the second `ABORT`, returns an empty list, and we don't update the state
- `snapshot` => returns `2` again
- `next_batch` => raises `ABORT` and ends the second run

The dataflow restarts with state `2`, while it should have been `3`, because we want to restart from the second `ABORT` itself.

The fix would be to advance the state if we see an abort with `item._triggered == True`. Doing this doesn't break any other test, and fixes this one, so I'm assuming it's just a bug, is that correct?